### ISSUE TITLE
MultiheadAttention incorrectly handles mask

### ISSUE THREAD

**Issue by yuwl798180** [pytorch/pytorch#27968]
## To Reproduce

```python
m = nn.MultiheadAttention(embed_dim=20, num_heads=5, dropout=0.1)

i = torch.randn(4, 5, 20)
q = k = v = i.transpose(0, 1)
key_padding_mask = seq_len_to_mask([5, 3, 1, 0], max_len=5)

ao, aow = m(q, k, v, key_padding_mask=key_padding_mask)
ao = ao.transpose(0, 1)

print(aow)
```

the print is:

         tensor([[[0.2822, 0.2024, 0.1926, 0.2565, 0.1463],
         [0.2020, 0.1906, 0.1721, 0.2159, 0.2425],
         [0.2074, 0.2127, 0.0882, 0.2422, 0.3333],
         [0.2298, 0.1938, 0.2093, 0.1432, 0.2335],
         [0.2531, 0.1772, 0.3045, 0.1246, 0.1881]],

        [[0.3693, 0.3857, 0.3561, 0.0000, 0.0000],
         [0.3655, 0.2712, 0.3918, 0.0000, 0.0000],
         [0.2230, 0.3795, 0.2549, 0.0000, 0.0000],
         [0.1698, 0.3805, 0.2318, 0.0000, 0.0000],
         [0.2986, 0.3093, 0.4405, 0.0000, 0.0000]],

        [[0.8889, 0.0000, 0.0000, 0.0000, 0.0000],
         [1.1111, 0.0000, 0.0000, 0.0000, 0.0000],
         [1.1111, 0.0000, 0.0000, 0.0000, 0.0000],
         [1.1111, 0.0000, 0.0000, 0.0000, 0.0000],
         [1.1111, 0.0000, 0.0000, 0.0000, 0.0000]],

        [[   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan]]], grad_fn=<DivBackward0>)

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):  1.3.0
 - OS (e.g., Linux):  macos
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

cc @ezyang @gchanan @zou3519

**Comment by yuwl798180** [comment]
when seq_len>1, and mask to 1, the attention weight cant correctly calculated.

**Comment by zou3519** [comment]
cc @zhangguanheng66

**Comment by zou3519** [comment]
This seems high pri because it is a correctness issue.

**Comment by yuwl798180** [comment]
and hopely add batch first api, because some nlp researchers really like putting batch as the first dim.

just see the above code.

**Comment by yuwl798180** [comment]
and hopely add attention_mask api, for at training time prune some heads to see what the specific head learning

by the way, can i modify the code and commit pr?

**Comment by gchanan** [comment]
@yuwl798180: sure, we generally accept PRs.  @zhangguanheng66 would probably have more guidance but he's out currently.

**Comment by ezyang** [comment]
@yuwl798180 What implementation of `seq_len_to_mask` are you using? Is this an accurate reproduction of your problem?

```
from torch import nn
import torch

def seq_len_to_mask(seq_len, max_len=None):
    batch_size = seq_len.size(0)
    max_len = int(max_len) if max_len else seq_len.max().long()
    broad_cast_seq_len = torch.arange(max_len).expand(batch_size, -1).to(seq_len)
    mask = broad_cast_seq_len.lt(seq_len.unsqueeze(1))
    return mask

m = nn.MultiheadAttention(embed_dim=20, num_heads=5, dropout=0.1)

i = torch.randn(4, 5, 20)
q = k = v = i.transpose(0, 1)
key_padding_mask = seq_len_to_mask(torch.tensor([5, 3, 1, 0]), max_len=5)

ao, aow = m(q, k, v, key_padding_mask=key_padding_mask)
ao = ao.transpose(0, 1)

print(aow)
```

Running this I get:

```
python test.py                               
tensor([[[   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan]],

        [[0.0000, 0.0000, 0.0000, 0.4940, 0.6172],
         [0.0000, 0.0000, 0.0000, 0.4223, 0.6031],
         [0.0000, 0.0000, 0.0000, 0.5560, 0.5551],
         [0.0000, 0.0000, 0.0000, 0.4385, 0.4451],
         [0.0000, 0.0000, 0.0000, 0.3753, 0.6067]],

        [[0.0000, 0.2734, 0.1930, 0.2423, 0.1207],
         [0.0000, 0.2145, 0.2673, 0.1718, 0.2661],
         [0.0000, 0.1328, 0.2039, 0.3825, 0.2168],
         [0.0000, 0.2458, 0.2853, 0.2572, 0.2635],
         [0.0000, 0.2411, 0.2660, 0.2597, 0.2615]],

        [[0.1679, 0.2105, 0.2043, 0.2306, 0.1731],
         [0.1757, 0.2134, 0.1370, 0.2089, 0.2518],
         [0.1703, 0.1278, 0.1588, 0.3244, 0.2705],
         [0.1856, 0.1483, 0.3194, 0.2495, 0.1608],
         [0.1221, 0.2183, 0.1858, 0.1405, 0.2774]]], grad_fn=<DivBackward0>)
```

**Comment by yuwl798180** [comment]
@ezyang  your seq_len_to_mask function is not equal with me.
the key_padding_mask key means for mask the pos that is padding.
so my func is just reverse with you, just print below:

```python
def seq_len_to_mask(seq_len, max_len=None, mask_pos_to_true=True):

    if isinstance(seq_len, list):
        seq_len = np.array(seq_len)

    if isinstance(seq_len, np.ndarray):
        seq_len = torch.from_numpy(seq_len)

    if isinstance(seq_len, torch.Tensor):
        assert seq_len.dim() == 1, f"seq_len can only have one dimension, got {seq_len.dim() == 1}."
        batch_size = seq_len.size(0)
        max_len = int(max_len) if max_len else seq_len.max().long()
        broad_cast_seq_len = torch.arange(max_len).expand(batch_size, -1).to(seq_len)
        if mask_pos_to_true:
            mask = broad_cast_seq_len.ge(seq_len.unsqueeze(1))
        else:
            mask = broad_cast_seq_len.lt(seq_len.unsqueeze(1))
    else:
        raise TypeError("Only support 1-d list or 1-d numpy.ndarray or 1-d torch.Tensor.")

    return mask

```

and reproducing my results,  you can just set the length is: [5, 4, 1, 0]

`key_padding_mask = seq_len_to_mask(torch.tensor([5, 4, 1, 0]), max_len=5)`

and print:

```
tensor([[[0.2623, 0.1638, 0.1304, 0.2848, 0.2309],
         [0.2702, 0.2537, 0.1625, 0.2159, 0.1302],
         [0.1857, 0.1523, 0.2133, 0.2395, 0.1643],
         [0.2067, 0.1865, 0.1998, 0.1929, 0.2431],
         [0.1845, 0.2478, 0.1853, 0.1237, 0.2216]],

        [[0.5133, 0.1585, 0.2141, 0.1938, 0.0000],
         [0.2647, 0.2372, 0.3771, 0.2321, 0.0000],
         [0.3619, 0.2027, 0.2398, 0.2843, 0.0000],
         [0.2233, 0.3000, 0.3638, 0.1770, 0.0000],
         [0.2246, 0.1748, 0.2159, 0.2421, 0.0000]],

        [[0.8889, 0.0000, 0.0000, 0.0000, 0.0000],
         [1.1111, 0.0000, 0.0000, 0.0000, 0.0000],
         [1.1111, 0.0000, 0.0000, 0.0000, 0.0000],
         [1.1111, 0.0000, 0.0000, 0.0000, 0.0000],
         [1.1111, 0.0000, 0.0000, 0.0000, 0.0000]],

        [[   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan],
         [   nan,    nan,    nan,    nan,    nan]]], grad_fn=<DivBackward0>)

```

**Comment by ezyang** [comment]
OK thanks!

**Comment by jerryzh168** [comment]
this seems to be a dup of another issue

**Comment by zhangguanheng66** [comment]
This is a duplicate issue as https://github.com/pytorch/pytorch/issues/24816.

Since you are padding a full sequence, a tensor of zeros is passed to the `softmax` func within `MultiheadAttention` func. We have discussed the results of `NaN` and it makes sense for the `softmax` func (https://github.com/pytorch/pytorch/issues/25110).
